### PR TITLE
[ruff] implement useless if-else (RUF034)

### PR DESCRIPTION
**Pull request by iamlucasvieira**
## Summary
Adds a new rule, RUF102, to check for redundant ternary operators. Message shows when both the true and false branches return the same value, as discussed in #12516

## Test Plan

`cargo test`

### CODE REVIEW

**Comment by RubenVanEldik** [comment]
Hi Lucas,

I am new to Ruff an am just reading PRs to get a better understanding of how everything works. I was wondering, how did you decide to name it `RUF102`. Is there any logic behind the `RUF0XX`, `RUF1XX` sets?

File: crates/ruff_linter/src/codes.rs [hunk 1]
```diff
@@ -963,6 +963,7 @@ pub fn code_to_rule(linter: Linter, code: &str) -> Option<(RuleGroup, Rule)> {
         (Ruff, "033") => (RuleGroup::Preview, rules::ruff::rules::PostInitDefault),
         (Ruff, "100") => (RuleGroup::Stable, rules::ruff::rules::UnusedNOQA),
         (Ruff, "101") => (RuleGroup::Stable, rules::ruff::rules::RedirectedNOQA),
+        (Ruff, "102") => (RuleGroup::Preview, rules::ruff::rules::RedundantTernary),
```
Comment (MichaReiser) [inline crates/ruff_linter/src/codes.rs:966]: The `1XX` ruff rules are reserved for noqa specific rules. Let's change the code to 034.

File: crates/ruff_linter/src/checkers/ast/analyze/expression.rs [hunk 2]
```diff
@@ -1404,6 +1404,9 @@ pub(crate) fn expression(expr: &Expr, checker: &mut Checker) {
             if checker.enabled(Rule::IfExpInsteadOfOrOperator) {
                 refurb::rules::if_exp_instead_of_or_operator(checker, if_exp);
             }
+            if checker.enabled(Rule::RedundantTernary) {
```
Comment (MichaReiser) [inline crates/ruff_linter/src/checkers/ast/analyze/expression.rs:1407]: I quickly checked other "ternary" specific rules and all of them use the term `if-expr` instead. We also use the term `Useless` when referring to *useless* code.

How about: `useless-if-expr`?

**Review by MichaReiser — COMMENTED** [review]
This overall looks great to me but we should change the rule code and its name.

I have some concerns with the scope of the rule. I know it has previously been accepted but I only just now looked closer into it. 

* There's some overlap with `SIM108` that simplifies if expressions
* The scope of the rule is very narrow. Should it also apply to if-statements?

It would be great if the rule also provided a fix that removes the if expression with the true or false body.

**Comment by MichaReiser** [comment]
The ecosystem check shows an interesting false positive:

https://github.com/apache/superset/blob/548d543efe81ecd6f0a6657550230b765ab4d955/superset/utils/decorators.py#L158

File: crates/ruff_linter/src/codes.rs (hunk 1 above)
Comment (iamlucasvieira) [reply crates/ruff_linter/src/codes.rs:966]: @RubenVanEldik I think this answers it 😄. This is also my first time contributing here.

**Comment by AlexWaygood** [comment]
> The ecosystem check shows an interesting false positive:
> 
> [apache/superset@`548d543`/superset/utils/decorators.py#L158](https://github.com/apache/superset/blob/548d543efe81ecd6f0a6657550230b765ab4d955/superset/utils/decorators.py#L158)

_Is_ that a false positive? The snippet in question is:

```py
sorted_args = tuple(
    x if hasattr(x, "__repr__") else x for x in [*args, *sorted(kwargs.items())]
)
```

It seems like each element in the iterable there will be collected whether or not it has a `__repr__` method, so I think it can just be simplified to

```py
sorted_args = tuple(x for x in [*args, *sorted(kwargs.items())])
```

or even just

```py
sorted_args = (*args, *sorted(kwargs.items()))
```

What they probably _wanted_ to do was

```py
sorted_args = tuple(
    repr(x) if hasattr(x, "__repr__") else x for x in [*args, *sorted(kwargs.items())]
)
```

But even that doesn't make much sense, because all objects in Python have a `__repr__` attribute:

```pycon
>>> hasattr(object, "__repr__")
True
```

File: crates/ruff_linter/src/checkers/ast/analyze/expression.rs (hunk 2 above)
Comment (AlexWaygood) [reply crates/ruff_linter/src/checkers/ast/analyze/expression.rs:1407]: `uesless-if-expr` sounds good to me. I agree that that's more approachable language than `ternary`

**Comment by AlexWaygood** [comment]
> There's some overlap with `SIM108` that simplifies if expressions

I support this being a separate rule to `SIM108`. `SIM108` is quite opinionated IMO: I think it's questionable whether rewriting an `if/else` block as a ternary expression is always easier to read in Python, or always more idiomatic. This rule, on the other hand, seems like it's pointing out something that you'd almost always want to avoid. A useless ternary expression is almost certainly a straight-up mistake that you didn't mean to write 😄

> The scope of the rule is very narrow. Should it also apply to if-statements?

Also applying it to if/else statements sounds reasonable to me.

**Comment by iamlucasvieira** [comment]
> It would be great if the rule also provided a fix that removes the if expression with the true or false body.

I don't think a fix would fit this rule because users probably end up in this situation when they mistype one of the ternary ends. Instead of just removing the ternary, they probably need to correct one of the sides.

File: crates/ruff_linter/src/checkers/ast/analyze/expression.rs (hunk 2 above)
Comment (iamlucasvieira) [reply crates/ruff_linter/src/checkers/ast/analyze/expression.rs:1407]: Sounds good, but if we also cover if blocks maybe we will need a different name. For example: `useless-if-logic`

Comment (AlexWaygood) [reply crates/ruff_linter/src/checkers/ast/analyze/expression.rs:1407]: hmm, good point. Maybe `useless-if-else`?

**Comment by iamlucasvieira** [comment]
> * The scope of the rule is very narrow. Should it also apply to if-statements?

I wonder if rule SIM114 address that

File: crates/ruff_linter/src/checkers/ast/analyze/expression.rs (hunk 2 above)
Comment (iamlucasvieira) [reply crates/ruff_linter/src/checkers/ast/analyze/expression.rs:1407]: I like`useless-if-else`

Comment (MichaReiser) [reply crates/ruff_linter/src/checkers/ast/analyze/expression.rs:1407]: I like `useless-if-else`. It provides room for expanding the rule in the future.

**Comment by MichaReiser** [comment]
`SIM114` only seems to flag `if-elif` but not `if-else`. https://play.ruff.rs/ae79ec83-f045-4262-bcfb-dc8a08a19460

**Review by AlexWaygood — APPROVED** [review]
LGTM. I think we could merge this now and expand it to `if`/`else` statements as a followup. @MichaReiser what do you think?

**Pull request merged**